From f18feb6f7c4c5cfafbba15b6d37f5a756a86f464 Mon Sep 17 00:00:00 2001
From: "Pandey, Amit" <Amit.Pandey@amd.com>
Date: Fri, 10 Oct 2025 09:14:01 +0530
Subject: [PATCH 9/9] [Compiler-RT] Improve hsa and comgr dependency search for
 AMDGPU ASan. (#4419)

> Add cmake explicit validation logic of 'hsa.h' header found by
find_path command.

> Remove ```CMAKE_INSTALL_PREFIX``` from `find_path` hints.
---
 compiler-rt/CMakeLists.txt                     | 18 +++++++++++++-----
 .../sanitizer_symbolizer_amdgpu.h              | 11 ++++-------
 2 files changed, 17 insertions(+), 12 deletions(-)

diff --git a/compiler-rt/CMakeLists.txt b/compiler-rt/CMakeLists.txt
index 098bbc021251..d8aef6e3797b 100644
--- a/compiler-rt/CMakeLists.txt
+++ b/compiler-rt/CMakeLists.txt
@@ -508,13 +508,21 @@ endif()
 
 if(SANITIZER_AMDGPU)
   list(APPEND SANITIZER_COMMON_CFLAGS -DSANITIZER_AMDGPU=1)
-  find_path(HSA_INCLUDE hsa.h HINTS ${SANITIZER_HSA_INCLUDE_PATH} ${CMAKE_INSTALL_PREFIX}/include/hsa ${CMAKE_INSTALL_PREFIX}/../include/hsa /opt/rocm/include/hsa PATH_SUFFIXES hsa)
+  message(STATUS "Looking 'hsa' and 'amd_comgr' header")
+  find_path(HSA_INCLUDE NAMES hsa.h HINTS ${SANITIZER_HSA_INCLUDE_PATH} /opt/rocm/include PATH_SUFFIXES hsa)
+  if(NOT HSA_INCLUDE)
+    message(FATAL_ERROR "Required header 'hsa.h' not found in path ${HSA_INCLUDE}. Aborting SANITIZER_AMDGPU build")
+  endif()
+  message(STATUS "Found 'hsa.h' in ${HSA_INCLUDE}")
   include_directories(${HSA_INCLUDE})
-  find_path(COMgr_INCLUDE amd_comgr.h.in HINTS ${SANITIZER_COMGR_INCLUDE_PATH} ${CMAKE_INSTALL_PREFIX}/../include ${CMAKE_INSTALL_PREFIX}/include)
-  if (NOT COMgr_INCLUDE)
-    add_compile_definitions(AMD_COMGR)
-    find_path(COMgr_INCLUDE amd_comgr.h HINTS ${CMAKE_INSTALL_PREFIX}/../include ${CMAKE_INSTALL_PREFIX}/include /opt/rocm/include)
+  find_path(COMgr_INCLUDE NAMES amd_comgr.h.in HINTS ${SANITIZER_COMGR_INCLUDE_PATH} PATH_SUFFIXES amd_comgr)
+  if(NOT COMgr_INCLUDE)
+    find_path(COMgr_INCLUDE NAMES amd_comgr.h HINTS /opt/rocm/include PATH_SUFFIXES amd_comgr)
+    if(NOT COMgr_INCLUDE)
+      message(FATAL_ERROR "Required header 'amd_comgr.h/amd_comgr.h.in' not found in path ${COMgr_INCLUDE}. Aborting SANITIZER_AMDGPU build")
+    endif()
   endif()
+  message(STATUS "Found 'amd_comgr.h.in/amd_comgr.h' in ${COMgr_INCLUDE}")
   include_directories(${COMgr_INCLUDE})
 endif()
 
diff --git a/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_amdgpu.h b/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_amdgpu.h
index 6a7aa93dac9c..196804a8af1f 100644
--- a/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_amdgpu.h
+++ b/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_amdgpu.h
@@ -13,15 +13,12 @@
 #if SANITIZER_AMDGPU
 #  include "sanitizer_common.h"
 #  include "sanitizer_symbolizer_internal.h"
-#  ifdef AMD_COMGR
+#  if __has_include("amd_comgr.h.in")
+#    include "amd_comgr.h.in"
+#  elif __has_include("amd_comgr.h")
 #    include "amd_comgr.h"
 #  else
-// Note: We require amd_comgr.h.in in COMgr source directory to be valid C
-// header file since we have a circular dependency between compiler-rt and COMgr
-// builds.
-// FIXME: Find a long-term solution for resolving this circular dependency and
-// avoid including amd_comgr.h.in.
-#    include "amd_comgr.h.in"
+#    error "No amd_comgr.h/amd_comgr header found!"
 #  endif
 
 namespace __sanitizer {
-- 
2.43.0

